File: backend/src/integrity/trust_labels.py
"""V2 trust-label contract — single source of truth for trust_label values.

Three values locked for V2 (see ``V2_DESIGN_SPEC.md §10`` and plan
``§3 Trust-label contract``):

- ``BENCHMARKED_CANONICAL_TEMPLATE`` — carried on the flagship
  ``agent_instances`` row deployed in Task 18 / D-1. The flagship ``agents``
  row carries ``subject_type = canonical_template`` and does NOT have a
  ``trust_label`` field.
- ``BENCHMARK_COMPATIBLE_CUSTOMIZED_INSTANCE`` — assigned to every
  user-deployed hosted instance when the deploy saga transitions to
  ``live``. Default value on the ``agent_instances.trust_label`` column.
- ``EXTERNAL_CUSTOM_RUNTIME`` — **reserved** in the V2 contract. No V2
  code path assigns it; the enum member exists so the schema, API
  responses, and the frontend recognise the value without a retrofit when
  (post-V2) an external-runtime path ships.

The DB-level CHECK constraint on ``agent_instances.trust_label`` rejects
any off-contract value. Because ``db.models`` cannot import from the
``src.integrity`` package (circular via its eager ``__init__``), the CHECK
list is duplicated as a tuple in ``db.models`` and held in sync with this
enum by a drift-guard test.
"""

from __future__ import annotations

from enum import Enum


class TrustLabel(str, Enum):
    """The three V2 trust-label values. See module docstring for semantics."""

    BENCHMARKED_CANONICAL_TEMPLATE = "benchmarked_canonical_template"
    BENCHMARK_COMPATIBLE_CUSTOMIZED_INSTANCE = (
        "benchmark_compatible_customized_instance"
    )
    EXTERNAL_CUSTOM_RUNTIME = "external_custom_runtime"


def trust_label_values() -> tuple[str, ...]:
    """Return the full set of valid ``agent_instances.trust_label`` strings.

    Consumed by callers building CHECK SQL or API responses — keeps them
    from hardcoding the vocabulary in multiple places.
    """

    return tuple(m.value for m in TrustLabel)

File: frontend/src/lib/trustLabels.ts
/**
 * V2 trust-label contract — single source of truth for the three trust-label
 * display strings rendered across flagship, template detail, and instance
 * dashboard surfaces.
 *
 * Keys are byte-equal to the backend ``TrustLabel`` enum values in
 * ``backend/src/integrity/trust_labels.py`` and the DB CHECK constraint on
 * ``agent_instances.trust_label``.
 *
 * Semantics:
 * - ``benchmarked_canonical_template``: carried on the flagship hosted
 *   instance.
 * - ``benchmark_compatible_customized_instance``: user-deployed hosted
 *   instance that completed the deploy saga.
 * - ``external_custom_runtime``: **reserved** in V2. No V2 code path
 *   assigns it; the value exists so UI renders consistently when the
 *   post-V2 external-runtime path ships.
 */

export const TRUST_LABELS = {
  benchmarked_canonical_template: 'Canonical Template',
  benchmark_compatible_customized_instance: 'Customized Instance',
  external_custom_runtime: 'External Runtime',
} as const;

/** Union of the three valid V2 trust-label string literals. */
export type TrustLabel = keyof typeof TRUST_LABELS;

/** Display name for a known trust label. */
export function getTrustLabelDisplay(label: TrustLabel): string {
  return TRUST_LABELS[label];
}

/** Type guard: narrow an arbitrary string to a known trust label.
 *
 * Uses ``Object.prototype.hasOwnProperty`` so inherited prototype keys like
 * ``toString`` / ``constructor`` are not mis-classified as valid labels.
 */
export function isTrustLabel(value: string): value is TrustLabel {
  return Object.prototype.hasOwnProperty.call(TRUST_LABELS, value);
}
